flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java

```diff
@@ -0,0 +1,187 @@
+            domainResolver.setAccessible(true);
+            return (ProtectionDomain[]) domainResolver.invoke(context);
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    void testProtectionDomainLeaksWithRegularFileSystemFactory() throws Exception {
+        testProtectionDomainLeak(
+                new TestFileSystemFactory() {
+
+                    @Override
+                    public FileSystem create(URI fsUri) {
+                        boolean leak = false;
+                        for (ProtectionDomain protectionDomain : getProtectionDomains()) {
+                            if (protectionDomain.getClassLoader()
+                                    instanceof ChildFirstClassLoader) {
+                                leak = true;
+                                break;
+                            }
+                        }
+                        assertThat(leak).isTrue();
+                        return null;
+                    }
+                });
+    }
+
+    @Test
+    void testProtectionDomainDoesNotLeakWithWrappedFileSystemFactory() throws Exception {
+        testProtectionDomainLeak(
+                SafeFileSystemFactory.of(
+                        new TestFileSystemFactory() {
+
+                            @Override
+                            public FileSystem create(URI fsUri) {
+                                for (ProtectionDomain protectionDomain : getProtectionDomains()) {
+                                    assertThat(protectionDomain.getCodeSource())
+                                            .isNotInstanceOf(ChildFirstClassLoader.class);
+                                }
+                                return null;
+                            }
+                        }));
+    }
+
+    private void testProtectionDomainLeak(FileSystemFactory factory) throws Exception {
+        final URL testClassLocation =
+                getClass().getProtectionDomain().getCodeSource().getLocation();
+        final ChildFirstClassLoader leakyClassLoader =
+                new ChildFirstClassLoader(
+                        new URL[] {testClassLocation},
+                        Thread.currentThread().getContextClassLoader(),
+                        new String[] {"java."},
+                        error -> {
+                            // No-op.
+                        });
+        final Class<?> clazz =
+                Class.forName(ProtectionDomainLeakTest.class.getName(), true, leakyClassLoader);
+        final Constructor<?> declaredConstructor =
+                clazz.getDeclaredConstructor(FileSystemFactory.class);
+        declaredConstructor.setAccessible(true);
+        declaredConstructor.newInstance(factory);
+        // We have to assert class name here, because the class we want to assert has been loaded in
+        // a different classloader.
```

Comment:
Initially, I didn't understand this comment here. Not sure whether it would help others as well mentioning the implicit fact that the `Class` object attached to an instance of the class is actually a new `Class` instance when being loaded from a different classloader